File: bc-w3/bcw3/Pen&Paper/Pen.cpp
#include <iostream>
#include "Pen.h"
#include "Paper.h"

Pen::Pen(int inkCapacity) : inkCapacity(inkCapacity), inkAmount(inkCapacity) {}

Pen::~Pen() {}

int Pen::getInkAmount() const {
    return inkAmount;
}

int Pen::getInkCapacity() const {
    return inkCapacity;
}

void Pen::write(Paper& paper, const std::string& message) {
    if ( this->inkAmount == 0 ) {
        throw OutOfInk();
    }
    
    int beforePaperSymbols = paper.getSymbols();
    if ( message.length() > this->inkAmount ) {
        paper.addContent(message.substr(0, this->inkAmount));
        this->inkAmount -= paper.getSymbols() - beforePaperSymbols;
        return;
    }
    paper.addContent(message);
    this->inkAmount -= paper.getSymbols() - beforePaperSymbols;
}

void Pen::refill() {
    inkAmount = inkCapacity;
}

File: bc-w3/java/Pen&Paper/Pen.java
public class Pen {
    private int inkAmount;
    private int inkCapacity;
    
    public Pen(int inkCapacity) {
        this.inkAmount = inkCapacity;
        this.inkCapacity = inkCapacity;
    }
    
    public Pen() {
        this(4096);
    }
    
    public int getInkAmount() {
        return inkAmount;
    }
    
    public int getInkCapacity() {
        return inkCapacity;
    }
    
    public void write(Paper paper, String message) throws OutOfInkException, OutOfSpaceException {
        if ( this.inkAmount == 0 ) {
            throw new OutOfInkException();
        }
        
        int beforePaperSymbols = paper.getSymbols();
        if ( message.length() > this.inkAmount ) {
            paper.addContent(message.substring(0, this.inkAmount));
            this.inkAmount -= paper.getSymbols() - beforePaperSymbols;
            return;
        }
        paper.addContent(message);
        this.inkAmount -= paper.getSymbols() - beforePaperSymbols;
    }
    
    public void refill() {
        inkAmount = inkCapacity;
    }
}
